solidity/ecdsa/test/WalletRegistry.Upgrade.test.ts

```diff
@@ -0,0 +1,336 @@
+import { deployments, ethers, upgrades, helpers } from "hardhat"
+import chai, { expect } from "chai"
+import chaiAsPromised from "chai-as-promised"
+import { keccak256 } from "ethers/lib/utils"
+
+import { params, walletRegistryFixture } from "./fixtures"
+import { fakeRandomBeacon } from "./utils/randomBeacon"
+import { noMisbehaved, signAndSubmitCorrectDkgResult } from "./utils/dkg"
+import ecdsaData from "./data/ecdsa"
+import { createNewWallet } from "./utils/wallets"
+
+import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import type { WalletRegistry, WalletRegistryV2 } from "../typechain"
+import type { FactoryOptions } from "hardhat/types"
+import type { Contract } from "ethers"
+import type { UpgradeProxyOptions } from "@openzeppelin/hardhat-upgrades/src/utils/options"
+
+const { mineBlocksTo } = helpers.time
+const { AddressZero } = ethers.constants
+
+chai.use(chaiAsPromised)
+
+describe("WalletRegistry - Upgrade", async () => {
+  let proxyAdminOwner: SignerWithAddress
+  let EcdsaInactivity: Contract
+
+  before(async () => {
+    proxyAdminOwner = await ethers.getNamedSigner("esdm")
+    EcdsaInactivity = await ethers.getContract("EcdsaInactivity")
+  })
+
+  describe("upgradeProxy", () => {
+    describe("when new contract fails upgradeability validation", () => {
+      describe("when a variable was added before old variables", () => {
+        it("should throw an error", async () => {
+          await deployments.fixture()
+
+          await expect(
+            upgradeProxy("WalletRegistry", "WalletRegistryV2Invalid", {
+              factoryOpts: {
+                libraries: { EcdsaInactivity: EcdsaInactivity.address },
+                signer: proxyAdminOwner,
+              },
+              proxyOpts: {
+                constructorArgs: [AddressZero, AddressZero],
+                unsafeAllowLinkedLibraries: true,
```

Comment:
If we decide to change to `unsafeAllow` (see my comment to `04_deploy_wallet_registry.ts`), then let's also remember to update this file. It's best to have the calls 1:1, just in case.

Reply:
8ef9325a39f175182163c51d3d43feec8bf0948f